File: hello/爬虫/动态渲染页面爬取/Splash/splash_.lua
function main(splash, args)
    local treat = require("treat")
    local json = require("json")
    local response = splash:http_post {
        "http://httpbin.org/post",
        body = json.encode({ name = "Germey" }),
        headers = { ["content-type"] = "application/json" }
    }
    -- 获取网页加载的过程描述
    local har = splash:har()
    return {
        har = har,
        html = treat.as_string(response.body),
        url = response.url,
        status = response.status
    }
end



--[[
    元素选择
--]]
function main(splash)
    splash:go("https://www.baidu.com/")
    -- css选择器
    input = splash:select("#kw")
    input:send_text('Splash')
    splash:wait(3)
    return splash:png()
end

function main(splash)
    local treat = require('treat')
    assert(splash:go("http://quotes.toscrape.com/"))
    -- 选择所有
    local texts = splash:select_all('.quote .text')
    local results = {}
    for index, text in ipairs(texts) do
        results[index] = text.node.innerHTML
    end
    return treat.as_array(results)
end


--[[
    模拟点击
-- ]]

function main(splash)
    splash:go("https://www.baidu.com/")
    input = splash:select("#kw")
    input:send_text('Splash')
    submit = splash:select('#su')
    submit:mouse_click()
    splash:wait(3)
    return splash:png()
end

File: hello/爬虫/动态渲染页面爬取/Splash/Splash_.py
# Splash是一个JavaScript渲染服务，是一个带有HTTP API的轻量级浏览器

"""
异步方式处理多个网页渲染过程；
获取渲染后的页面的源代码或截图；
通过关闭图片渲染或者使用Adblock规则来加快页面渲染速度；
可执行特定的JavaScript脚本；
可通过Lua脚本来控制页面渲染过程；
获取渲染的详细过程并通过HAR（HTTP Archive）格式呈现


Splash 是通过Lue脚本控制的页面加载过程，模拟浏览器
"""

import requests
# url = 'http://localhost:8050/render.html?url=https://www.baidu.com'
# response = requests.get(url)
# print(response.text)
#
# url = 'http://localhost:8050/render.png?url=https://www.jd.com&wait=5&width=1000&height=700'
# response = requests.get(url)
# with open('jd.png', 'wb') as f:
#     f.write(response.content)


# 和Lua交互
from urllib.parse import quote

lua = '''
function main(splash, args)
  local treat = require("treat")
  local response = splash:http_get("http://httpbin.org/get")
    return {
    html=treat.as_string(response.body),
    url=response.url,
    status=response.status
    }
end
'''
# quote URL转码
url = 'http://localhost:8050/execute?lua_source=' + quote(lua)
resonse = requests.get(url)
print(resonse.text)
